### PR TITLE
[trainer] feat: add support for the GDPO algorithm

### PR DESCRIPTION
**Pull request by yue-zeng-yue**
## Description
This PR introduces support for the GDPO algorithm into the training framework. It enables the system to handle multiple reward functions effectively, which is a key requirement for multi-objective reinforcement learning tasks.

## Reference Paper & Experimental Results
GDPO Paper: https://arxiv.org/abs/2601.05242

Performance Benchmarks:

Tool Calling: GDPO increases the correct format ratio from 76.33% to 80.66% and improves overall accuracy from 30.18% to 32.81% (BFCL-v3).

Math Reasoning: On AIME (DeepSeek-R1-7B), GDPO improves accuracy from 50.2% to 53.1% and reduces length-exceeding violations from 2.1% to 0.2%.

Coding Reasoning: In CodeContests (3-objective optimization), GDPO reduces the bug ratio significantly from 13.2% to 3.9% compared to GRPO.

## Changes Made
Core Algorithm: Updated verl/trainer/ppo/core_algos.py to implement GDPO advantage computation logic.

Ray Trainer: Modified verl/trainer/ppo/ray_trainer.py to support multi-reward data preparation and metrics logging.

Configuration: Added necessary parameters in verl/trainer/config/algorithm.py to enable GDPO-specific settings.

## Motivation
The current framework's limitations made it difficult to natively implement algorithms like GDPO that require complex or multiple reward function handling. This update modifies the core PPO trainer and Ray trainer to fully support GDPO.

### CODE REVIEW

**Comment by tardis-key** [comment]
there are multi scores in gdpo rewards , and will be used in advEstimator laster. The current implementation does not seem to align with this algorithm.

File: verl/trainer/ppo/ray_trainer.py [hunk 1]
```diff
@@ -838,9 +857,8 @@ def init_workers(self):
             reward_loop_worker_handles=reward_loop_worker_handles,
         )
 
-        checkpoint_engine_config = omega_conf_to_dataclass(self.config.actor_rollout_ref.rollout.checkpoint_engine)
         self.checkpoint_manager = CheckpointEngineManager(
-            config=checkpoint_engine_config,
+            backend=self.config.actor_rollout_ref.rollout.checkpoint_engine.backend,
```
Comment (wuxibin89) [inline verl/trainer/ppo/ray_trainer.py]: Please rebase main first, this should not be changed.

**Comment by yue-zeng-yue** [comment]
> there are multi scores in gdpo rewards , and will be used in advEstimator laster. The current implementation does not seem to align with this algorithm.

Thank you for the review! I’ve double-checked the code, and it seems to align with the underlying GDPO calculation logic. Could you please point out specifically which part doesn't match the algorithm? I’d appreciate more details so I can fix it properly.

File: verl/trainer/ppo/ray_trainer.py (hunk 1 above)
Comment (yue-zeng-yue) [reply verl/trainer/ppo/ray_trainer.py]: ok，got it

**Comment by tardis-key** [comment]
In verl/trainer/ppo/core_algos.py: line 395:
reward_components: (bs, N_rewards) – per-sample scores for each reward dimension.

These sample-level rewards do not take into account prompt_mask or attention_mask, resulting in significant differences compared to the official GDPO implementation, in [GDPO/verl-GDPO/verl/trainer/main_ppo.py](https://github.com/NVlabs/GDPO/blob/main/verl-GDPO/verl/trainer/main_ppo.py): line 85
`score, fomrat_score, correctness_score, length_score = compute_score_fn(solution_str=sequences_str, ground_truth=ground_truth, step=step)`

you can refer to https://github.com/verl-project/verl/pull/5422, which adapted reward_score

**Comment by yue-zeng-yue** [comment]
> In verl/trainer/ppo/core_algos.py: line 395: reward_components: (bs, N_rewards) – per-sample scores for each reward dimension.
> 
> These sample-level rewards do not take into account prompt_mask or attention_mask, resulting in significant differences compared to the official GDPO implementation, in [GDPO/verl-GDPO/verl/trainer/main_ppo.py](https://github.com/NVlabs/GDPO/blob/main/verl-GDPO/verl/trainer/main_ppo.py): line 85 `score, fomrat_score, correctness_score, length_score = compute_score_fn(solution_str=sequences_str, ground_truth=ground_truth, step=step)`
> 
> you can refer to #5422, which adapted reward_score


> Thanks for the review and for referencing #5422.
>
> To clarify, my compute_gdpo_outcome_advantage in core_algos.py is the advantage computation step — it receives pre-computed per-dimension rewards as input and implements the three GDPO steps (group-wise decoupled normalization → weighted aggregation → batch-level normalization), which is mathematically consistent with the paper.
>
> I agree that my PR currently lacks the multi-reward scoring pipeline (reward manager + compute_score function) to produce those per-dimension rewards. My PR assumes they are already available in non_tensor_batch, and the GDPO-related fields in algorithm.py (i.e., gdpo_reward_keys and gdpo_reward_weights) only define configuration parameters without the actual reward computation logic.

**Comment by yue-zeng-yue** [comment]
> In verl/trainer/ppo/core_algos.py: line 395: reward_components: (bs, N_rewards) – per-sample scores for each reward dimension.
> 
> These sample-level rewards do not take into account prompt_mask or attention_mask, resulting in significant differences compared to the official GDPO implementation, in [GDPO/verl-GDPO/verl/trainer/main_ppo.py](https://github.com/NVlabs/GDPO/blob/main/verl-GDPO/verl/trainer/main_ppo.py): line 85 `score, fomrat_score, correctness_score, length_score = compute_score_fn(solution_str=sequences_str, ground_truth=ground_truth, step=step)`
> 
> you can refer to #5422, which adapted reward_score

> My design intention was to keep the advantage computation decoupled from any specific reward function, so that users can plug in their own compute_score that returns a dict with custom reward keys.

File: verl/trainer/ppo/ray_trainer.py [hunk 2]
```diff
@@ -199,6 +199,35 @@ def compute_advantage(
             adv_kwargs["index"] = data.non_tensor_batch["uid"]
         if "reward_baselines" in data.batch:  # optional
             adv_kwargs["reward_baselines"] = data.batch["reward_baselines"]
+        # GDPO: extract per-dimension reward components from non_tensor_batch
+        # and convert them to token-level tensors for compatibility with GRPO
+        if adv_estimator in (AdvantageEstimator.GDPO, "gdpo"):
```
Comment (wuxibin89) [inline verl/trainer/ppo/ray_trainer.py:204]: Please move #L204-#L230 into `compute_gdpo_outcome_advantage`
